src/models/Control.php

```diff
@@ -216,6 +216,92 @@ class Control extends Model {
     await Level::genBaseScoring();
   }
 
+  public static async function genAutoBegin(): Awaitable<void> {
+    // Get start time
+    $config_start_ts = await Configuration::gen('start_ts');
+    $start_ts = intval($config_start_ts->getValue());
+
+    // Get end time
+    $config_end_ts = await Configuration::gen('end_ts');
+    $end_ts = intval($config_end_ts->getValue());
+
+    // Get paused status
+    $config_game_paused = await Configuration::gen('game_paused');
+    $game_paused = intval($config_game_paused->getValue());
+
+    if (($game_paused === 0) && ($start_ts <= time()) && ($end_ts > time())) {
+      // Start the game
+      await Control::genBegin();
+    }
+  }
+
+  public static async function genAutoEnd(): Awaitable<void> {
+    // Get start time
+    $config_start_ts = await Configuration::gen('start_ts');
+    $start_ts = intval($config_start_ts->getValue());
+
+    // Get end time
+    $config_end_ts = await Configuration::gen('end_ts');
+    $end_ts = intval($config_end_ts->getValue());
+
+    // Get paused status
+    $config_game_paused = await Configuration::gen('game_paused');
+    $game_paused = intval($config_game_paused->getValue());
+
+    if (($game_paused === 0) && ($end_ts <= time())) {
+      // Start the game
+      await Control::genEnd();
+    }
+  }
+
+  public static async function genAutoRun(): Awaitable<void> {
+    // Get start time
+    $config_game = await Configuration::gen('game');
+    $game = intval($config_game->getValue());
+
+    if ($game === 0) {
+      // Check and start the game
+      await Control::genAutoBegin();
+    } else {
+      // Check and stop the game
+      await Control::genAutoEnd();
+    }
+  }
+
+  public static async function genRunAutoRunScript(): Awaitable<void> {
+    $autorun_status = await Control::checkScriptRunning('autorun');
+    if ($autorun_status === false) {
+      $document_root = must_have_string(Utils::getSERVER(), 'DOCUMENT_ROOT');
+      $cmd =
+        'hhvm -vRepo.Central.Path=/var/run/hhvm/.hhvm.hhbc_autorun '.
+        $document_root.
```

Comment:
@iangcarroll pointed out to me that if the user does a manual install and puts the document root in a non-standard location, it's possible that the document root could contain spaces and mess up this command. It's also possible (although not likely) that the document root could contain unsafe special characters.